### PR TITLE
Validation for `module.loaders` and friends

### PR DESCRIPTION
**Pull request by jonathanewerner**
This validates `module.loaders`, `module.preLoaders` & `module.postLoaders`.
I think two reviewers would be nice to have here, quite a lot of conditions and stuff in there. :)

### CODE REVIEW

File: src/validators/module/index.js [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import {noop} from 'lodash'
+import validateLoadersEntry from './validateLoadersEntry'
+
 export default [
-  {key: 'module.loaders', validate: noop},
-  {key: 'module.preLoaders', validate: noop},
-  {key: 'module.postLoaders', validate: noop},
+  {key: 'module.loaders', validate: validateLoadersEntry},
```
Comment (kentcdodds) [inline src/validators/module/index.js]: `Entry` doesn't make sense in this context. `validateLoaders` should do just fine

File: src/utils/friendly-type-of/index.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ const typesAndInstances = [
   {instance: Boolean, type: 'boolean'},
   {instance: Number, type: 'number'},
   {instance: Array, type: 'array'},
+  {instance: Function, type: 'function'},
```
Comment (kentcdodds) [inline src/utils/friendly-type-of/index.js]: :+1:

**Comment by kentcdodds** [comment]
Wow! Lots of stuff. Will finish reviewing later

File: src/validators/module/index.js (hunk 1 above)
Comment (jonathanewerner) [inline src/validators/module/index.js]: :+1: Wanted to conform to `validateEntry` function name convention we have in the codebase. But `validateLoaders` is clearer, agreed. Refactored it.

File: src/validators/module/validateLoader.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  if (loader && loaders) {
+    messages.push(`Found \`loader\` and \`loaders\`. These properties are mutually exclusive`)
+  }
+
+  if (loaders && query) {
+    messages.push(`You can only pass the \`query\` property when you have one \`loader\``)
+  }
+
+  if (loader && friendlyTypeOf(loader) !== 'string') {
+    messages.push(`Expected \`loader\` property to be a string, but got ${JSON.stringify(loader)}`)
```
Comment (kentcdodds) [inline src/validators/module/validateLoader.js]: Probably just do the friendly type of here rather than JSON. Thoughts? Just worry it could be big...

Comment (jonathanewerner) [inline src/validators/module/validateLoader.js]: Well a loader object JSON won't be that big. The longest could be the loader string, maybe 80-100 chars. But I'm a fan of failing verbosely and I think it'll be an UX benefit to see which loader is faulty at first sight.

Comment (kentcdodds) [inline src/validators/module/validateLoader.js]: :+1: cool

File: src/validators/module/validateLoader.test.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+})
+
+test('passes with valid `test` & `loaders` & `exclude` & `include`', t => {
+  const validLoader = {loaders: ['foo!bar', 'baz'], test: /foo/, include: /foo/, exclude: /baz/}
+  const result = validateLoader(validLoader)
+  t.notOk(result)
+})
+
+test('passes with an arbitrary real world loader', t => {
+  const realWorldLoader = {
```
Comment (kentcdodds) [inline src/validators/module/validateLoader.test.js]: :+1:

**Comment by kentcdodds** [comment]
Looks awesome

**Pull request merged**